### ISSUE TITLE
"From ... Require" doesn't detect ambiguous logical names

### ISSUE THREAD

**Issue by jfehrle** [coq/coq#16314]
Unlike `Require`, `From ... Require` doesn't report ambiguous paths.  It appears to the first item found in the loadpath.

For example:

ambig/a.v = `Ltac amb := idtac "ambig".`
ambig/ambig2/a.v = `Ltac amb := idtac "ambig2".`

Compile:
`ambig$ dune exec -- coqc a.v -R . Ambig`
`ambig/ambig2$ dune exec -- coqc a.v -R . Ambig.ambig2`

Run `$ dune exec -- dev/shim/coqide-prelude -R ambig Ambig`.
`Print LoadPath.` shows:

```
Logical Path / Physical path:
Ambig /home/proj/coq2/ambig
Ambig.ambig2 /home/proj/coq2/ambig/ambig2
  :
```

Then

```
From Ambig Require Import a.
Goal True.
amb.  (* prints "ambig" *)
```

while

```
Require Import a.  (* Required library a matches several files in path (found /home/proj/coq2/ambig/a.vos and
                      /home/proj/coq2/ambig/ambig2/a.vos). *)
```

@herbelin Is this your area of expertise?

**Comment by Alizter** [comment]
I think even if you swapped the order of the loadpath it would still choose `ambig`. It picks the "shortest" name resolution in a sense. `From Ambig Require a` means `Require A.*a` in a sense, and Coq has to find something to fill it, preferring a shorter one.

`Require a`, on the other hand, really is ambiguous since it is not obvious which it should prefer.

**Comment by jfehrle** [comment]
Well, I'm sure one can describe what it actually does, but the point is that a) it's unfriendly behavior and b) inconsistent with `Require`.  BTW, no way to swap the two paths if they're in the same package.

**Comment by herbelin** [comment]
Indeed, there is a preference for "exact matches" (see https://coq.inria.fr/refman/proof-engine/vernacular-commands.html?highlight=require#coq:cmd.From-%E2%80%A6-Require).

In which sense do you find it unfriendly? Do you mean that the rule is not intuitive enough?

Both `From Require` and `Require` apply the exact match rule. For instance, in the following example, there is no exact match for `From Require` and an ambiguity is reported:
```
ambig/ambig1/a.v = Ltac amb := idtac "ambig1".
ambig/ambig2/a.v = Ltac amb := idtac "ambig2".
```
compiled e.g. with
```
ambig/ambig1$ dune exec -- coqc a.v -R . Ambig.ambig1
ambig/ambig2$ dune exec -- coqc a.v -R . Ambig.ambig2
```
then, within:
```
ambig$ dune exec -- coqtop.byte -R . Ambig
```
the call to
```
From Ambig Require Import a.
```
reports an ambiguity (in the absence of an exact match).

**Comment by jfehrle** [comment]
Thanks for explaining.  I see it works because it's an exact match, which happens to be the shortest match for my example, but the notion of "shortest" doesn't apply in general.  It has to work that way--you can't add more qualification to an exact match!

FWIW, I had written some [text](https://github.com/coq/coq/pull/15888#pullrequestreview-1034464095) in March for #15888 saying that it was ambiguous (based on an earlier experiment) before later reading about the exact match rule, which didn't come to mind yesterday.
